extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,354 @@
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+
+public class KinesisSinkP<T> implements Processor {
+
+    //todo; Each shard can support writes up to 1,000 records per second, up to a
+    // maximum data write total of 1 MiB per second. Right now we don't pre-check
+    // this, because we don't check which item goes to what shard, we just rely
+    // failure handling with exponential backoff. Would be complicated to improve
+    // on, but can we affort not to?
+
+    /**
+     * PutRecords requests are limited to 500 records.
+     */
+    private static final int MAX_RECORDS_IN_REQUEST = 500;
+
+    private static final long PAUSE_AFTER_FAILURE = 1000L; //todo: exponential backoff
+
+    /**
+     * Each record, when encoded as a byte array, is limited to 1M,
+     * including the partition key (Unicode String).
+     */
+    private static final int MAX_RECORD_SIZE_IN_BYTES = 1024 * 1024;
+
+    /**
+     * The maximum allowed size of all the records in a PutRecords request,
+     * including partition keys is 5M.
+     */
+    private static final int MAX_REQUEST_SIZE_IN_BYTES = 5 * 1024 * 1024;
+
+    /**
+     * The number of Unicode characters making up keys is limited to a maximum
+     * of 256.
+     */
+    private static final int MAX_UNICODE_CHARS_IN_KEY = 256;
+
+    private final AmazonKinesisAsync kinesis;
+    private final String stream;
+
+    private final Buffer<T> buffer;
+
+    private ILogger logger;
+    private KinesisHelper helper;
+
+    private State state = State.READY_TO_SEND;
+    private long nextSendTime;
+    private Future<PutRecordsResult> sendResult;
+
+    public KinesisSinkP(
+            AmazonKinesisAsync kinesis,
+            @Nonnull String stream,
+            @Nonnull FunctionEx<T, String> keyFn,
+            @Nonnull FunctionEx<T, byte[]> valueFn
+    ) {
+        this.kinesis = kinesis;
+        this.stream = stream;
+        this.buffer = new Buffer<>(keyFn, valueFn);
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return true;
+    }
+
+    @Override
+    public void init(@Nonnull Outbox outbox, @Nonnull Context context) {
+        logger = context.logger();
+        helper = new KinesisHelper(kinesis, stream, logger);
+        helper.waitForStreamToActivate();
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true; //watermark ignored
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        switch (state) {
+            case READY_TO_SEND:
+                handleReadyToSend(inbox);
+                return;
+            case SENDING_IN_PROGRESS:
+                handleSendingInProgress();
+                return;
+            default:
+                throw new RuntimeException("Programming error, unhandled state: " + state);
+        }
+
```

Comment:
In the current code, if sending in progress was just done, we won't send another batch immediately. We'll wait for a next call. The ProcessorTasklet will think no progress was made, so it will back off a little, unnecessarily.
```suggestion
        if (state == State.SENDING_IN_PROGRESS) {
            handleSendingInProgress();
        }
        if (state == State.READY_TO_SEND) {
            handleReadyToSend(inbox);
        }
```
Note there's no `else` before the second `if` - the state might change after the first call.

Reply:
Makes sense, will apply.